### PR TITLE
Fall back to English depots for unavailable languages

### PR DESCRIPTION
**Pull request by VinceBT**
## Description
Some owned games show **0 B** download and disk size and can't be installed. Example: appId 1150690 on a **French** container.

The cause is depot language selection. That game ships one full Windows depot **per language** (English, Simplified Chinese, Japanese, Korean) with no language-neutral build. `filterForDownloadableDepots` drops every non-DLC depot whose `language` doesn't match the container language, and none of them is French, so **all depots are filtered out**: zero downloadable depots, hence 0 B and no install.

Fix: added `SteamUtils.effectiveDepotLanguage`, which picks the language to filter by from what the app actually ships an installable depot in — the requested one, else English (Steam's default), else any language it offers. Depot resolution filters by that effective language instead of the raw container language.

The helper lives in `SteamUtils` so both the base game (`resolveDownloadableDepots`) and each DLC (`getDownloadableDepots`) resolve their own language independently, so we don't pull wrong-language files for a DLC. It only pools tagged base-game depots when choosing the fallback, so an owned in-app DLC can't steer the base game's language, and untagged (neutral) depots pass the language filter unconditionally and are ignored by the pool, so they resolve regardless. Games that already offer the requested language are unaffected.

Result: the French container now resolves the English depot for 1150690 and shows a real size / can install. Any game shipping per-language depots without the user's language is fixed the same way, base game and DLC alike.

## Recording

| Before | After |
| :---: | :---: |
| Screenshot_20260710_141605_GameNative | Screenshot_20260710_141307_GameNative |

## Type of Change
- [x] Bug fix
- [ ] Performance / stability improvement
- [ ] Compatibility improvements
- [ ] Other (requires prior approval)

## Checklist
- [x] If I have access to `#code-changes`, I have discussed this change there and it has been green-lighted. If I do not have access, I have still provided clear context in this PR. If I skip both, I accept that this change may face delays in review, may not be reviewed at all, or may be closed.
- [x] This change aligns with the current project scope (core functionality, stability, or performance). If not, it has been explicitly approved beforehand.
- [x] I have attached a recording of the change.
- [x] I have read and agree to the contribution guidelines in `CONTRIBUTING.md`.

---
## Summary by cubic
Fall back to English (or another shipped language) when a game’s container language isn’t available in Windows depots, so base games and DLC show real sizes and install instead of 0 B.

- **Bug Fixes**
  - Resolve an effective depot language via `SteamUtils.effectiveDepotLanguage` from installable base‑game Windows depots: requested → `english` → first available; neutral depots don’t steer; DLC depots don’t steer the base game; arch ignored; exclude `SteamChina`; encrypted manifests count only with a Steam‑unlocked branch; respect licensing and DLC ownership.
  - Apply that language in eligibility and filtering for the main app and each DLC, fixing titles missing the container language; added unit tests for priorities and edge cases.

<sup>Written for commit 71681056d0a4ae44399c52b3bc4eaceba31fa557. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/utkarshdalal/GameNative/pull/1659?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved depot download language selection by deriving the most effective, currently installable base-game language rather than relying solely on the preferred language.
  * Enhanced fallback behavior to choose a suitable language when the preferred option isn’t available or eligible for download.
  * Updated DLC depot filtering so each DLC uses its own effective language, improving compatibility and reducing missing or failed DLC downloads.

### CODE REVIEW

File: app/src/main/java/app/gamenative/service/SteamService.kt [hunk 1]
```diff
@@ -932,11 +932,37 @@ class SteamService : Service(), IChallengeUrlChanged {
...
+                .filter { it.language.isNotEmpty() && it.installableInItsLanguage() }
+                .mapTo(mutableSetOf()) { it.language }
+
+            // Prefer the requested language, then English, then whatever the app offers. Falling back
+            // keeps a game installable when it does not ship the container's language.
+            val effectiveLanguage = when {
+                preferredLanguage in availableLanguages -> preferredLanguage
+                "english" in availableLanguages -> "english"
+                else -> availableLanguages.firstOrNull() ?: preferredLanguage
+            }
```
[the remark this answers is not in the document]
Comment (phobos665) [reply app/src/main/java/app/gamenative/service/SteamService.kt]: @VinceBT  Worth discussing this

[the remark this answers is not in the document]
Comment (VinceBT) [reply app/src/main/java/app/gamenative/service/SteamService.kt]: Good catch. Scoped the language pool to base-game depots (dlcAppId == INVALID_APP_ID) so an owned in-app DLC's language can't steer the base game. DLC still gets its own fallback via the per-DLC effectiveDepotLanguage call in the getDownloadableDepots loop.

File: app/src/main/java/app/gamenative/utils/SteamUtils.kt [hunk 2]
```diff
@@ -101,6 +101,43 @@ object SteamUtils {
...
+        // A depot installs once its language is chosen if it passes every check except language
+        // and arch. Arch is left out because it is a per-language preference, not a gate.
+        fun DepotInfo.installableInItsLanguage(): Boolean {
+            val isDlc = dlcAppId != SteamService.INVALID_APP_ID
+            val hasContent = manifests.isNotEmpty() || sharedInstall ||
+                (hasSteamUnlockedBranch && encryptedManifests.isNotEmpty())
+            val ownedIfDlc = !isDlc || ownedDlc == null || ownedDlc.containsKey(depotId)
+            val licensedIfBaseGame = isDlc || systemDefined ||
+                licensedDepotIds == null || depotId in licensedDepotIds
+            return isWindowsCompatible && hasContent && ownedIfDlc && licensedIfBaseGame
```
[the remark this answers is not in the document]
Comment (VinceBT) [reply app/src/main/java/app/gamenative/utils/SteamUtils.kt]: Good catch. Mirrored the SteamChina realm gate in installableInItsLanguage so the fallback can't pick a language the final filter later drops, and added a regression test.

**Comment by utkarshdalal** [comment]
Please add back the PR template, it should not be overwritten

**Comment by VinceBT** [comment]
Updated @utkarshdalal

**Review by phobos665 — COMMENTED** [review]
Before I give approval, could you please post screenshots of before & after? Also, are you now able to play those games? 

Would be good for end-result proof before merge

**Comment by VinceBT** [comment]
@phobos665 It's good

**Comment by utkarshdalal** [comment]
@joshuatam could you take a look as well please?

**Comment by joshuatam** [comment]
@VinceBT Looks good to me, just have a question.

Do you have Witcher 3 (292030)?
I wonder if it is possible to skip depot 292032 when the container language is French (which should only download 292038)
Currently, with & without your changes merged in, GN will still download both language depot due to 292032 doesn't have a language data.

```
SteamService$Companion   I  selectedDepots is empty? false
SteamService$Companion   I  Starting download for 292030
SteamService$Companion   I  App contains 12 depot(s): [228984, 228985, 228988, 228990, 292031, 292032, 292034, 292036, 292038, 292042, 292033, 370009]
SteamService$Companion   I  DLC contains 0 depot(s): []
SteamService$Companion   I  downloadingAppIds: [292030]
```

**Comment by joshuatam** [comment]
> @VinceBT Looks good to me, just have a question.
> 
> Do you have Witcher 3 (292030)? I wonder if it is possible to skip depot 292032 when the container language is French (which should only download 292038) Currently, with & without your changes merged in, GN will still download both language depot due to 292032 doesn't have a language data.
> 
> ```
> SteamService$Companion   I  selectedDepots is empty? false
> SteamService$Companion   I  Starting download for 292030
> SteamService$Companion   I  App contains 12 depot(s): [228984, 228985, 228988, 228990, 292031, 292032, 292034, 292036, 292038, 292042, 292033, 370009]
> SteamService$Companion   I  DLC contains 0 depot(s): []
> SteamService$Companion   I  downloadingAppIds: [292030]
> ```

Not possible, double checked PC Steam, english speech depot will still download even the game language is set to simplified chinese.

@utkarshdalal Good to merge for this PR.
